### PR TITLE
Technologies and service creation docs page

### PR DESCRIPTION
**Pull request by legendofa**
This is an overview page that should help when creating a new service. The page also describes the tech stack and its role in the architecture.

### CODE REVIEW

File: docs/dev-manuals/tooling/techstack-and-service-creation.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+This is a general overview of some of the technologies used in MiSArch and which services / components they involve.
+
+<CustomImage path="/images/techstack-overview" width="2803" height="2341" />
+
+A user client sends a request to Nginx to retrieve the frontend. The frontend then communicates with the backend through Nginx. Nginx acts both as a web-server and reverse proxy. The frontend creates requests in GraphQL. The GraphQL gateway resolves the request by querying the respective services.
+
+Dapr acts as a middleware abstracting communication between services. This enables metric and trace collection via Open Telemetry. Services communicate via Dapr sidecars. Each sidecar is associated with one service and known by the Dapr placement / master service, also known as Dapr Operator Service. The Dapr Operator feeds the Open Telemetry collector with metrics and traces.
+
+A full Zipkin stack stores and displays traces, while metrics are collected by Grafana / Prometheus. Each Service involved in GraphQL queries needs to expose a GraphQL compatible API. Each services needs to be registered in the GraphQL gateway.
```
Comment (nk-coding) [inline docs/dev-manuals/tooling/techstack-and-service-creation.md]: - each service involved in GraphQL queries needs to expose a GraphQL federation compatible API
- no, only services involved in GraphQL queries need to be registered in the gateway

File: docs/dev-manuals/tooling/techstack-and-service-creation.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+### Infastructure (as Code)
+
+The current plan is to support two deployments:
+
+1. Kubernetes
+
+   - State-of-the-art container orchestration. Widely used for microservices in science and the industry. Powerful but tedious to set up. Supports multi-node deployments, key stores, load balancing, auto scaling, etc.
```
Comment (nk-coding) [inline docs/dev-manuals/tooling/techstack-and-service-creation.md]: Note sure if I agree on the tedious one, in particular with terraform we could basically reach a one command deployment

Comment (nk-coding) [reply docs/dev-manuals/tooling/techstack-and-service-creation.md]: but I think we could call this one production ready (more or less)

File: docs/dev-manuals/tooling/techstack-and-service-creation.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+2. Docker Compose
+   - Single-node deployments. Good portability + easy setup. Declarative multi-container app tool. Packaged with Docker
+
+**Why two deployments?**
+
+Docker Compose assists development by enabling faster code + deployment iterations than Kubernetes. Kubernetes is a requirement and therefore MiSArch needs to be deployed with Kubernetes from time-to-time. We do not want to big-bang port MiSArch to Kubernetes at the end of the project. Incremental ports help to detect deployment difficulties early on.
+
+**(as Code)?**
+
+We plan on using Terraform for Kubernetes to deploy in a declarative way, which enables to version our deployments according to the infrastructure as code paradigm. Docker Compose might also benefit from Terraform as it enables configuration manipulation while abstracting boiler-plate code. This can for example reduce port mapping conflicts.
```
Comment (nk-coding) [inline docs/dev-manuals/tooling/techstack-and-service-creation.md]: did you do some reasearch if docker compose is reasonably supported with terraform? I'm not sure

File: static/images/techstack-overview-light.svg
Comment (nk-coding) [inline static/images/techstack-overview-light.svg:1]: are you sure about the dapr sidecars for the databases?
I've not seen that yet, and that could also make deployment via helm difficult there

File: docs/dev-manuals/tooling/techstack-and-service-creation.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+- GraphQL Federation 2 support: https://www.apollographql.com/docs/federation/building-supergraphs/supported-subgraphs/
+
+### Writing a new service step by step
+
+1. Expose services GraphQL API
+1. Integrate the service with Docker Compose & Kubernetes
+   1. Containerize the service
+   1. Set up service container Dapr sidecar
+   1. (Set up database)
+   1. (Set up database Dapr sidecar)
```
Comment (nk-coding) [inline docs/dev-manuals/tooling/techstack-and-service-creation.md]: are you sure about that one?

**Review by nk-coding — CHANGES_REQUESTED** [review]

File: docs/dev-manuals/tooling/techstack-and-service-creation.md (hunk 3 above)
Comment (legendofa) [reply docs/dev-manuals/tooling/techstack-and-service-creation.md]: It looks like Terraform supports Docker, but not Docker Compose https://developer.hashicorp.com/terraform/tutorials/docker-get-started/docker-build. Makes sense as Docker Compose is already declarative on its own. The question is if we gain advantages by using Terraform + Docker in comparison to Docker Compose.

Also, if Terraform makes it less tedious by being declarative, why not only support Terraform + Kubernetes and drop Docker Compose?

File: static/images/techstack-overview-light.svg
Comment (legendofa) [reply static/images/techstack-overview-light.svg:1]: Makes sense, databases should not need sidecars, as their are not managed with Dapr

File: docs/dev-manuals/tooling/techstack-and-service-creation.md [hunk 5]
```diff
@@ -16,15 +16,15 @@ A user client sends a request to Nginx to retrieve the frontend. The frontend th
...
+A full Zipkin stack stores and displays traces, while metrics are collected by Grafana / Prometheus. Each Service involved in GraphQL queries needs to expose a GraphQL Federation 2 compatible API. Services involved in GraphQL queries need to be registered in the GraphQL gateway.
 
 ### Infastructure (as Code)
 
 The current plan is to support two deployments:
 
 1. Kubernetes
 
-   - State-of-the-art container orchestration. Widely used for microservices in science and the industry. Powerful but tedious to set up. Supports multi-node deployments, key stores, load balancing, auto scaling, etc.
+   - State-of-the-art container orchestration. Widely used for microservices in science and the industry. Powerful but tedious to set up (without infrastructure as code tools). Supports multi-node deployments, key stores, load balancing, auto scaling, etc.
```
Comment (nk-coding) [inline docs/dev-manuals/tooling/techstack-and-service-creation.md]: but we use infrastructure as code tools
idk, mentioning sth is tedious without what is basically the title of the section is a bit strange to me
but idk

File: docs/dev-manuals/tooling/techstack-and-service-creation.md (hunk 3 above)
Comment (nk-coding) [reply docs/dev-manuals/tooling/techstack-and-service-creation.md]: we decided in call to support both, as docker compose is helpful for developing and also easier to deploy for reasearchers when trying the reference architecture

**Review by nk-coding — APPROVED** [review]

File: docs/dev-manuals/tooling/techstack-and-service-creation.md [hunk 6]
```diff
@@ -0,0 +1,62 @@
+---
+sidebar_position: 3
+---
+
+import CustomImage from "@site/src/components/CustomImage";
+
+# Techstack and service creation
+
+### Techstack pverview
```
Comment (lesser1) [inline docs/dev-manuals/tooling/techstack-and-service-creation.md]: Typo?! Should be Overview, right?

File: static/images/techstack-overview-light.svg
Comment (lesser1) [inline static/images/techstack-overview-light.svg:1]: The two versions -- light and dark -- are different, aren't they? I think, in the light one, there are some of the icons of the manufacturers missing, Idk.

**Review by lesser1 — DISMISSED** [review]
The typo should be fixed imo. Other than that, I certainly approve this PR. The readability of the page is really good! This is of great value also to future stakeholders.

File: static/images/techstack-overview-light.svg
Comment (legendofa) [reply static/images/techstack-overview-light.svg:1]: Ah yes, the images are missing and need to be included

**Comment by delvh** [comment]
@lesser1 Same procedure as with #22.

File: docs/dev-manuals/tooling/techstack-and-service-creation.md [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+This is a general overview of some of the technologies used in MiSArch and which services / components they involve.
+
+<CustomImage path="/images/techstack-overview" width="2803" height="2341" />
+
+A user client sends a request to Nginx to retrieve the frontend. The frontend then communicates with the backend through Nginx. Nginx acts both as a web-server and reverse proxy. The frontend creates requests in GraphQL. The GraphQL gateway resolves the request by querying the respective services.
+
+Dapr acts as a middleware abstracting communication between services. This enables metric and trace collection via Open Telemetry. Services communicate via Dapr sidecars. Each sidecar is associated with one service and known by the Dapr placement / master service, also known as Dapr Operator Service. The Dapr Operator feeds the Open Telemetry collector with metrics and traces.
+
+A full Zipkin stack stores and displays traces, while metrics are collected by Grafana / Prometheus. Each Service involved in GraphQL queries needs to expose a GraphQL Federation 2 compatible API. Services involved in GraphQL queries need to be registered in the GraphQL gateway.
```
Comment (delvh) [inline docs/dev-manuals/tooling/techstack-and-service-creation.md]: I think we should add a link here to what we mean by that.
Judging by this sentence alone, I have no idea what is required of each service.

File: docs/dev-manuals/tooling/techstack-and-service-creation.md [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+We plan on using Terraform for Kubernetes to deploy in a declarative way, which enables to version our deployments according to the infrastructure as code paradigm.
+
+# Creating a service
+
+A developer needs to take the following steps when writing a new service for MiSArch.
+This is due to the extensive tech-stack that the project supports.
+
+Services can be written in the language of choice if the languages ecosystem supports:
+
+- GraphQL Federation 2 support: https://www.apollographql.com/docs/federation/building-supergraphs/supported-subgraphs/
```
Comment (delvh) [inline docs/dev-manuals/tooling/techstack-and-service-creation.md]: Why is that sentence only down here?

File: docs/dev-manuals/tooling/techstack-and-service-creation.md [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+### Writing a new service step by step
```
Comment (delvh) [inline docs/dev-manuals/tooling/techstack-and-service-creation.md:47]: Shouldn't this be its own document?

**Review by lesser1 — APPROVED** [review]

File: docs/dev-manuals/tooling/techstack-and-service-creation.md (hunk 9 above)
Comment (legendofa) [reply docs/dev-manuals/tooling/techstack-and-service-creation.md:47]: Could be its own document, but i will leave it for now. We can extract it when someone writes a more detailed guide

File: docs/dev-manuals/tooling/techstack-and-service-creation.md (hunk 7 above)
Comment (legendofa) [reply docs/dev-manuals/tooling/techstack-and-service-creation.md]: I have added the link here as well

File: docs/dev-manuals/tooling/techstack-and-service-creation.md (hunk 5 above)
Comment (legendofa) [reply docs/dev-manuals/tooling/techstack-and-service-creation.md]: Ok, i have removed that

**Review by delvh — APPROVED** [review]

**Pull request merged**